In [7]:
# Import required libraries and dependencies
import pandas as pd
import hvplot.pandas
from sklearn.cluster import KMeans
from sklearn.decomposition import PCA
from sklearn.preprocessing import StandardScaler

In [8]:
# Load the data into a Pandas DataFrame
df_market_data = pd.read_csv(
    "Resources/crypto_market_data.csv",
    index_col="coin_id")

# Display sample data
df_market_data.head(10)

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                    
bitcoin                              1.08388                     7.60278   
ethereum                             0.22392                    10.38134   
tether                              -0.21173                     0.04935   
ripple                              -0.37819                    -0.60926   
bitcoin-cash                         2.90585                    17.09717   
binancecoin                          2.10423                    12.85511   
chainlink                           -0.23935                    20.69459   
cardano                              0.00322                    13.99302   
litecoin                            -0.06341                     6.60221   
bitcoin-cash-sv                      0.92530                     3.29641   

                 price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                     
bitcoin                              6.57509                      7.67258   
ethereum                             4.80849                      0.13169   
tether                               0.00640                     -0.04237   
ripple                               2.24984                      0.23455   
bitcoin-cash                        14.75334                     15.74903   
binancecoin                          6.80688                      0.05865   
chainlink                            9.30098                    -11.21747   
cardano                              5.55476                     10.10553   
litecoin                             7.28931                      1.21662   
bitcoin-cash-sv                     -1.86656                      2.88926   

                 price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                      
bitcoin                             -3.25185                      83.51840   
ethereum                           -12.88890                     186.77418   
tether                               0.28037                      -0.00542   
ripple                             -17.55245                      39.53888   
bitcoin-cash                       -13.71793                      21.66042   
binancecoin                         36.33486                     155.61937   
chainlink                          -43.69522                     403.22917   
cardano                            -22.84776                     264.51418   
litecoin                           -17.23960                      27.49919   
bitcoin-cash-sv                    -24.87434                       7.42562   

                 price_change_percentage_1y  
coin_id                                      
bitcoin                            37.51761  
ethereum                          101.96023  
tether                              0.01954  
ripple                            -16.60193  
bitcoin-cash                       14.49384  
binancecoin                        69.69195  
chainlink                         325.13186  
cardano                           156.09756  
litecoin                          -12.66408  
bitcoin-cash-sv                    93.73082

In [9]:
# Generate summary statistics
df_market_data.describe()

price_change_percentage_24h  price_change_percentage_7d  \
count                    41.000000                   41.000000   
mean                     -0.269686                    4.497147   
std                       2.694793                    6.375218   
min                     -13.527860                   -6.094560   
25%                      -0.608970                    0.047260   
50%                      -0.063410                    3.296410   
75%                       0.612090                    7.602780   
max                       4.840330                   20.694590   

       price_change_percentage_14d  price_change_percentage_30d  \
count                    41.000000                    41.000000   
mean                      0.185787                     1.545693   
std                       8.376939                    26.344218   
min                     -18.158900                   -34.705480   
25%                      -5.026620                   -10.438470   
50%                       0.109740                    -0.042370   
75%                       5.510740                     4.578130   
max                      24.239190                   140.795700   

       price_change_percentage_60d  price_change_percentage_200d  \
count                    41.000000                     41.000000   
mean                     -0.094119                    236.537432   
std                      47.365803                    435.225304   
min                     -44.822480                     -0.392100   
25%                     -25.907990                     21.660420   
50%                      -7.544550                     83.905200   
75%                       0.657260                    216.177610   
max                     223.064370                   2227.927820   

       price_change_percentage_1y  
count                   41.000000  
mean                   347.667956  
std                   1247.842884  
min                    -17.567530  
25%                      0.406170  
50%                     69.691950  
75%                    168.372510  
max                   7852.089700

In [10]:
# Plot your data to see what's in your DataFrame
df_market_data.hvplot.line(
    width=800,
    height=400,
    rot=90
)

:NdOverlay   [Variable]
   :Curve   [coin_id]   (value)

---

### Prepare the Data

In [15]:
# Use the `StandardScaler()` module from scikit-learn to normalize the data from the CSV file
df_scaled = StandardScaler().fit_transform(df_market_data[["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"]])

In [16]:
# Create a DataFrame with the scaled data
df_market_data_scaled = pd.DataFrame(df_scaled, columns=["price_change_percentage_24h", "price_change_percentage_7d", "price_change_percentage_14d", "price_change_percentage_30d", "price_change_percentage_60d", "price_change_percentage_200d", "price_change_percentage_1y"])

# Copy the crypto names from the original data
df_market_data_scaled["coin_id"] = df_market_data.index

# Set the coinid column as index
df_market_data_scaled = df_market_data_scaled.set_index("coin_id")

# Display sample data
df_market_data_scaled.head()

price_change_percentage_24h  price_change_percentage_7d  \
coin_id                                                                 
bitcoin                          0.508529                    0.493193   
ethereum                         0.185446                    0.934445   
tether                           0.021774                   -0.706337   
ripple                          -0.040764                   -0.810928   
bitcoin-cash                     1.193036                    2.000959   

              price_change_percentage_14d  price_change_percentage_30d  \
coin_id                                                                  
bitcoin                          0.772200                     0.235460   
ethereum                         0.558692                    -0.054341   
tether                          -0.021680                    -0.061030   
ripple                           0.249458                    -0.050388   
bitcoin-cash                     1.760610                     0.545842   

              price_change_percentage_60d  price_change_percentage_200d  \
coin_id                                                                   
bitcoin                         -0.067495                     -0.355953   
ethereum                        -0.273483                     -0.115759   
tether                           0.008005                     -0.550247   
ripple                          -0.373164                     -0.458259   
bitcoin-cash                    -0.291203                     -0.499848   

              price_change_percentage_1y  
coin_id                                   
bitcoin                        -0.251637  
ethereum                       -0.199352  
tether                         -0.282061  
ripple                         -0.295546  
bitcoin-cash                   -0.270317

---

### Find the Best Value for k Using the Original Data.

In [17]:
# Create a list with the number of k-values from 1 to 11
k = list(range(1, 11))

In [ ]:
# Create an empty list to store the inertia values
inertia = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_market_data_scaled`
# 3. Append the model.inertia_ to the inertia list
for i in k:
    k_model = KMeans(n_clusters=i, random_state=0)
    k_model.fit(df_market_data_scaled)
    inertia.append(k_model.inertia_)

In [31]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data = {"k": k, "inertia": inertia}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow = pd.DataFrame(elbow_data)
df_elbow

k     inertia
0   1  287.000000
1   2  195.820218
2   3  123.190482
3   4   79.022435
4   5   65.302379
5   6   52.888518
6   7   43.914690
7   8   37.517032
8   9   32.485241
9  10   28.222899

In [194]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot_kmeans = df_elbow.hvplot.line(
                    x="k", 
                    y="inertia", 
                    title="Elbow Curve", 
                    xticks=k
                    )

elbow_plot_kmeans

:Curve   [k]   (inertia)

#### Answer the following question: 

**Question:** What is the best value for `k`?

**Answer:** 4

---

### Cluster Cryptocurrencies with K-means Using the Original Data

In [25]:
# Initialize the K-Means model using the best value for k
model = KMeans(n_clusters=4, random_state=0)

In [ ]:
# Fit the K-Means model using the scaled data
model.fit(df_market_data_scaled)

In [27]:
# Predict the clusters to group the cryptocurrencies using the scaled data
kmeans_predictions = model.predict(df_market_data_scaled)

# Print the resulting array of cluster values.
kmeans_predictions

array([0, 0, 3, 3, 0, 0, 0, 0, 0, 3, 3, 3, 3, 0, 3, 0, 3, 3, 0, 3, 3, 0,
       3, 3, 3, 3, 3, 3, 0, 3, 3, 3, 1, 0, 3, 3, 2, 3, 3, 3, 3])

In [28]:
# Create a copy of the DataFrame
kmeans_predictions_df = df_market_data_scaled.copy()

In [ ]:
# Add a new column to the DataFrame with the predicted clusters
kmeans_predictions_df['crypto group'] = kmeans_predictions

# Display sample data
kmeans_predictions_df.head(20)

In [178]:
# Create a scatter plot using hvPlot by setting 
# `x="price_change_percentage_24h"` and `y="price_change_percentage_7d"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
kmeans_predict_plot = kmeans_predictions_df.hvplot.scatter(
                    x="price_change_percentage_24h", 
                    y="price_change_percentage_7d", 
                    by="crypto group",
                    hover_cols=["coin_id"]
                    ).opts(bgcolor="beige", text_color="purple", border_line_color="brown")

kmeans_predict_plot

:NdOverlay   [crypto group]
   :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)

---

### Optimize Clusters with Principal Component Analysis.

In [65]:
# Create a PCA model instance and set `n_components=3`.
pca = PCA(n_components=3)

In [80]:
# Use the PCA model with `fit_transform` to reduce to 
# three principal components.
crypto_pca_data = pca.fit_transform(df_market_data_scaled)

# View the first five rows of the DataFrame. 
crypto_pca_data[:5]

array([[-0.60066733,  0.84276006,  0.46159457],
       [-0.45826071,  0.45846566,  0.95287678],
       [-0.43306981, -0.16812638, -0.64175193],
       [-0.47183495, -0.22266008, -0.47905316],
       [-1.15779997,  2.04120919,  1.85971527]])

In [81]:
# Retrieve the explained variance to determine how much information 
# can be attributed to each principal component.
pca.explained_variance_ratio_

array([0.3719856 , 0.34700813, 0.17603793])

In [82]:
# Calculating the total explained variance for the three principal components
pca_evr = pca.explained_variance_ratio_
total_explained_variance = sum(pca_evr)

# Print the total explained variance of the three principal components
print(total_explained_variance)
print("The total explained variance of the three principal components is: {:.7f}".format(total_explained_variance))

0.895031657030984
The total explained variance of the three principal components is: 0.8950317


#### Answer the following question: 

**Question:** What is the total explained variance of the three principal components?

**Answer:** The total explained variance of the three principal components is: 0.8950317.

In [84]:
# Create a new DataFrame with the PCA data.
df_cryptos_pca = pd.DataFrame(crypto_pca_data, columns=["PC1", "PC2", "PC3"])

# Copy the crypto names from the original data
df_cryptos_pca["coin_id"] = df_market_data.index

# Set the coin_id column as index
df_cryptos_pca = df_cryptos_pca.set_index("coin_id")

# Display sample data
df_cryptos_pca.head() 

PC1       PC2       PC3
coin_id                                   
bitcoin      -0.600667  0.842760  0.461595
ethereum     -0.458261  0.458466  0.952877
tether       -0.433070 -0.168126 -0.641752
ripple       -0.471835 -0.222660 -0.479053
bitcoin-cash -1.157800  2.041209  1.859715

---

### Find the Best Value for k Using the PCA Data

In [98]:
# Create a list with the number of k-values from 1 to 11
k_pca = list(range(1, 11))

In [ ]:
# Create an empty list to store the inertia values
inertia_pca = []

# Create a for loop to compute the inertia with each possible value of k
# Inside the loop:
# 1. Create a KMeans model using the loop counter for the n_clusters
# 2. Fit the model to the data using `df_cryptos_pca`
# 3. Append the model.inertia_ to the inertia list
for i in k_pca:
    model = KMeans(n_clusters=i, random_state=0)
    model.fit(df_cryptos_pca)
    inertia_pca.append(model.inertia_)

In [101]:
# Create a dictionary with the data to plot the Elbow curve
elbow_data_pca = {
    "k_pca": k_pca,
    "inertia_pca": inertia_pca
}

# Create a DataFrame with the data to plot the Elbow curve
df_elbow_pca = pd.DataFrame(elbow_data_pca)
df_elbow_pca.head()

k_pca  inertia_pca
0      1   256.874086
1      2   165.901994
2      3    93.774626
3      4    49.665497
4      5    37.878747

In [103]:
# Plot a line chart with all the inertia values computed with 
# the different values of k to visually identify the optimal value for k.
elbow_plot_pca = df_elbow_pca.hvplot.line(x="k_pca", y="inertia_pca", title="Elbow Curve Using PCA Data", xticks=k_pca)
elbow_plot_pca

:Curve   [k_pca]   (inertia_pca)

#### Answer the following questions: 

* **Question:** What is the best value for `k` when using the PCA data?

  * **Answer:** 4


* **Question:** Does it differ from the best k value found using the original data?

  * **Answer:** Based on this Elbow Curve, it looks like k=4 is still the correct one.

### Cluster Cryptocurrencies with K-means Using the PCA Data

In [112]:
# Initialize the K-Means model using the best value for k
model2 = KMeans(n_clusters=4, random_state=0)

In [ ]:
# Fit the K-Means model using the PCA data
model2.fit(df_cryptos_pca)

In [114]:
# Predict the clusters to group the cryptocurrencies using the PCA data
crypto_clusters = model2.predict(df_cryptos_pca)

# Print the resulting array of cluster values.
print(crypto_clusters)

[0 0 3 3 0 0 0 0 0 3 3 3 3 0 3 0 3 3 0 3 3 0 3 3 3 3 3 3 0 3 3 3 1 0 3 3 2
 3 3 3 3]


In [115]:
# Create a copy of the DataFrame with the PCA data
df_cryptos_pca_predictions = df_cryptos_pca.copy()

# Add a new column to the DataFrame with the predicted clusters
df_cryptos_pca_predictions["CryptoCluster"] = crypto_clusters

# Display sample data
df_cryptos_pca_predictions.head()

PC1       PC2       PC3  CryptoCluster
coin_id                                                  
bitcoin      -0.600667  0.842760  0.461595              0
ethereum     -0.458261  0.458466  0.952877              0
tether       -0.433070 -0.168126 -0.641752              3
ripple       -0.471835 -0.222660 -0.479053              3
bitcoin-cash -1.157800  2.041209  1.859715              0

In [163]:
# Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC2"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_scatter_1PC = df_cryptos_pca_predictions.hvplot.scatter(
    x="PC1", 
    y="PC2", 
    by="CryptoCluster",
    hover_cols=["coin_id"]
).opts(bgcolor="lightyellow", text_color="brown", border_line_color="pink")

pca_scatter_1PC

:NdOverlay   [CryptoCluster]
   :Scatter   [PC1]   (PC2,coin_id)

In [165]:
# Extra: Create a scatter plot using hvPlot by setting 
# `x="PC1"` and `y="PC3"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_scatter_2PC = df_cryptos_pca_predictions.hvplot.scatter(x='PC1', y='PC3',
              by="CryptoCluster",
              hover_cols=["coin_id"]
              ).opts(bgcolor="lightyellow", text_color="brown", border_line_color="pink")
pca_scatter_2PC

:NdOverlay   [CryptoCluster]
   :Scatter   [PC1]   (PC3,coin_id)

In [167]:
# Extra: Create a scatter plot using hvPlot by setting 
# `x="PC2"` and `y="PC3"`. 
# Color the graph points with the labels found using K-Means and 
# add the crypto name in the `hover_cols` parameter to identify 
# the cryptocurrency represented by each data point.
pca_scatter_3PC = df_cryptos_pca_predictions.hvplot.scatter(x='PC2', y='PC3',
              by="CryptoCluster",
              hover_cols=["coin_id"]
              ).opts(bgcolor="lightyellow", text_color="brown", border_line_color="pink")
pca_scatter_3PC

:NdOverlay   [CryptoCluster]
   :Scatter   [PC2]   (PC3,coin_id)

### Visualize and Compare the Results

In this section, you will visually analyze the cluster analysis results by contrasting the outcome with and without using the optimization techniques.

In [184]:
# Composite plot to contrast the Elbow curves (layout of separate plots)
elbow_plot_kmeans + elbow_plot_pca 

:Layout
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k_pca]   (inertia_pca)

In [196]:
# Extra: Composite plot (overlay) to contrast the Elbow curves
elbow_plot_kmeans * elbow_plot_pca

:Overlay
   .Curve.I  :Curve   [k]   (inertia)
   .Curve.II :Curve   [k_pca]   (inertia_pca)

In [197]:
# Composite plot to contrast the clusters
kmeans_predict_plot + pca_scatter_1PC

:Layout
   .NdOverlay.I  :NdOverlay   [crypto group]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [CryptoCluster]
      :Scatter   [PC1]   (PC2,coin_id)

In [198]:
# Extra: Composite plot to contrast the clusters
kmeans_predict_plot * pca_scatter_1PC

:Overlay
   .NdOverlay.I  :NdOverlay   [crypto group]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II :NdOverlay   [CryptoCluster]
      :Scatter   [PC1]   (PC2,coin_id)

In [179]:
# Extra: Composite plot to contrast the clusters 
# (for better visualization, plotting the charts side by side offers better understanding on the clusters distribution than plotting all of them at once like previoulsy)
kmeans_predict_plot + pca_scatter_1PC + pca_scatter_2PC + pca_scatter_3PC

:Layout
   .NdOverlay.I   :NdOverlay   [crypto group]
      :Scatter   [price_change_percentage_24h]   (price_change_percentage_7d,coin_id)
   .NdOverlay.II  :NdOverlay   [CryptoCluster]
      :Scatter   [PC1]   (PC2,coin_id)
   .NdOverlay.III :NdOverlay   [CryptoCluster]
      :Scatter   [PC1]   (PC3,coin_id)
   .NdOverlay.IV  :NdOverlay   [CryptoCluster]
      :Scatter   [PC2]   (PC3,coin_id)

#### Answer the following question: 

  * **Question:** After visually analyzing the cluster analysis results, what is the impact of using fewer features to cluster the data using K-Means?

  * **Answer:** With fewer features, the points in each cluster get closer to each other as the variances decrease, and it's easier to visually determine the different clusters so we can better see which points belong to each cluster.